packages/cletus/src/helpers/dba-query.ts

```diff
@@ -0,0 +1,1351 @@
+      }
+    } else {
+      rows = records.map((record) => {
+        const row: Record<string, unknown> = {};
+        for (const aliasValue of stmt.values) {
+          row[aliasValue.alias] = evaluateValue(aliasValue.value, record, ctx);
+        }
+        return row;
+      });
+    }
+  }
+
+  // Apply DISTINCT
+  if (stmt.distinct) {
+    const seen = new Set<string>();
+    rows = rows.filter((row) => {
+      const key = JSON.stringify(row);
+      if (seen.has(key)) return false;
+      seen.add(key);
+      return true;
+    });
+  }
+
+  // Apply ORDER BY
+  if (stmt.orderBy?.length) {
+    rows = sortRows(rows, stmt.orderBy, ctx);
+  }
+
+  // Apply OFFSET and LIMIT
+  if (stmt.offset !== undefined && stmt.offset !== null) {
+    rows = rows.slice(stmt.offset);
+  }
+  if (stmt.limit !== undefined && stmt.limit !== null) {
+    rows = rows.slice(0, stmt.limit);
+  }
+
+  return { rows };
+}
+
+/**
+ * Execute an INSERT statement
+ */
+async function executeInsert(
+  stmt: Insert,
+  ctx: QueryContext
+): Promise<QueryResult> {
+  const dataManager = await getDataManager(stmt.table, ctx);
+  const insertedIds: string[] = [];
+  const insertedRecords: DataRecord[] = [];
+
+  // Get values to insert
+  let valuesToInsert: Record<string, unknown>[];
+
+  if (stmt.values?.length) {
+    // Direct values
+    const record: Record<string, unknown> = {};
+    for (let i = 0; i < stmt.columns.length; i++) {
+      record[stmt.columns[i]] = evaluateValue(stmt.values[i], null, ctx);
+    }
+    valuesToInsert = [record];
+  } else if (stmt.select) {
+    // Values from SELECT
+    const selectResult = await executeStatement(stmt.select, ctx);
+    valuesToInsert = selectResult.rows.map((row) => {
+      const record: Record<string, unknown> = {};
+      for (let i = 0; i < stmt.columns.length; i++) {
+        const selectAlias = Object.keys(row)[i];
+        record[stmt.columns[i]] = row[selectAlias];
+      }
+      return record;
+    });
+  } else {
+    throw new Error("INSERT must have either values or select");
+  }
+
+  // Insert records
+  for (const fields of valuesToInsert) {
+    if (stmt.onConflict) {
+      // Check for conflict
+      const conflictKey = stmt.onConflict.columns
+        .map((col) => String(fields[col]))
+        .join("_");
+      const existing = dataManager.getAll().find((record) => {
+        const existingKey = stmt.onConflict!.columns
+          .map((col) => String(record.fields[col]))
+          .join("_");
+        return existingKey === conflictKey;
+      });
+
+      if (existing) {
+        if (stmt.onConflict.doNothing) {
+          continue;
+        } else if (stmt.onConflict.update?.length) {
+          // Update on conflict
+          const updates: Record<string, unknown> = {};
+          for (const cv of stmt.onConflict.update) {
+            updates[cv.column] = evaluateValue(cv.value, existing, ctx);
+          }
+          await dataManager.update(existing.id, updates);
```

Comment:
Push all dataManager calls to the very end - using one commit function that does all validations against the type definitions before it rights. That means convert all dataManager calls to something that adds ops to commit to the context. And instead of dataManager.getAll everywhere the ctx needs to store a transactional state of each source/table that is manipulated. So if we have a cte that inserts records into a table, then updates some records, then deletes some records - the final state of the txn will be the minimal number of inserts/updates/deletes. So we don't insert something that was deleted, or update something that is about to be inserted, etc.